src/scene/world-clusters.js

```diff
@@ -0,0 +1,663 @@
+import { Vec3 } from '../math/vec3.js';
+import { math } from '../math/math.js';
+import { BoundingBox } from '../shape/bounding-box.js';
+import { Texture } from '../graphics/texture.js';
+import { PIXELFORMAT_R8_G8_B8_A8, PIXELFORMAT_RGBA32F, ADDRESS_CLAMP_TO_EDGE, TEXTURETYPE_DEFAULT, FILTER_NEAREST } from '../graphics/constants.js';
+import { LIGHTTYPE_DIRECTIONAL, LIGHTTYPE_SPOT } from './constants.js';
+
+const tempVec3 = new Vec3();
+const tempMin3 = new Vec3();
+const tempMax3 = new Vec3();
+const tempBox = new BoundingBox();
+
+const epsilon = 0.000001;
+const oneDiv255 = 1 / 255;
+const maxTextureSize = 4096;    // maximum texture size allowed to work on all devices
+
+// packs a float value in [0..1) range to specified number of bytes and stores them in an array with start offset
+// based on: https://aras-p.info/blog/2009/07/30/encoding-floats-to-rgba-the-final/
+// Note: calls to Math.round are only needed on iOS, precision is somehow really bad without it,
+// looks like an issue with their implementation of Uint8ClampedArray
+function packFloatToBytes(value, array, offset, numBytes) {
+    const enc1 = (255.0 * value) % 1;
+    array[offset + 0] = Math.round(((value % 1) - oneDiv255 * enc1) * 255);
+
+    if (numBytes > 1) {
+        const enc2 = (65025.0 * value) % 1;
+        array[offset + 1] = Math.round((enc1 - oneDiv255 * enc2) * 255);
+
+        if (numBytes > 2) {
+            const enc3 = (16581375.0 * value) % 1;
+            array[offset + 2] = Math.round((enc2 - oneDiv255 * enc3) * 255);
+
+            if (numBytes > 3) {
+                array[offset + 3] = Math.round(enc3 * 255);
+            }
+        }
+    }
+}
+
+// helper class to store properties of a light used by clustering
+class ClusterLight {
+    constructor() {
+        // the light itself
+        this.light = null;
+
+        // bounding box
+        this.min = new Vec3();
+        this.max = new Vec3();
+    }
+}
+
+class WorldClusters {
+    // format for high precision light texture - float
+    static FORMAT_FLOAT = 0;
+
+    // format for high precision light texture - half
+    static FORMAT_HALF = 1;
```

Comment:
looks like this is unused?

Reply:
yep, removed for now, can be re-added later when half precision implementation is done